src/main/java/com/google/api/codegen/util/ConfigNextVersionValidator.java

```diff
@@ -0,0 +1,79 @@
+/* Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.api.codegen.util;
+
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.util.JsonFormat;
+import com.google.protobuf.util.JsonFormat.Parser;
+import com.google.protobuf.util.JsonFormat.Printer;
+import com.google.protobuf.util.JsonFormat.TypeRegistry;
+import javax.annotation.Nonnull;
+
+public class ConfigNextVersionValidator {
+
+  public static String CONFIG_V2_MAJOR_VERSION = "2";
+  public static String CONFIG_V2_VERSION = CONFIG_V2_MAJOR_VERSION + ".0.0"; // "2.0.0"
+
+  /**
+   * Throw {@link IllegalStateException} iff the given input contains fields unknown to the {@link
+   * com.google.api.codegen.v2.ConfigProto} schema.
+   */
+  public void checkIsNextVersionConfig(@Nonnull com.google.api.codegen.ConfigProto configV1Proto)
+      throws IllegalStateException {
+    if (!configV1Proto.getConfigSchemaVersion().startsWith(CONFIG_V2_MAJOR_VERSION + ".")
+        && !configV1Proto.getConfigSchemaVersion().equals(CONFIG_V2_MAJOR_VERSION)) {
+      throw new IllegalStateException(
+          String.format(
+              "Provided ConfigProto version is %s but should be >= %s",
+              configV1Proto.getConfigSchemaVersion(), CONFIG_V2_VERSION));
+    }
+
+    try {
+      // Serializing utils for Config V2.
+      TypeRegistry typeRegistryV2 =
```

Comment:
Is it necessary to use a TypeRegistry and json formatting functions? Can we not just use toByteArray, parseFrom(byte[]), and check for unknown fields?

Reply:
If we use the toByteArray and parseFrom functions, those will still [preserve the unknown fields](https://github.com/protocolbuffers/protobuf/issues/272#issuecomment-405729394). The thing with the unknown fields is that they can exist at every single level of the protobuf object tree, and I don't feel like writing a tree search function to parse through every child message type and looking for unknown types (i've already been burned trying to implement a BFS haha)

Reply:
What about using a [DiscardUnknownFieldsParser](https://github.com/protocolbuffers/protobuf/blob/master/java/core/src/main/java/com/google/protobuf/DiscardUnknownFieldsParser.java#L43)? I don't know, but my intuition is that there is an almost-one-liner that does this, something like:
```
configBytes = config.toByteArray();
configV2Bytes = DiscardUnknownFieldsParser.wrap(ConfigV2.parser())
  .parseFrom(configBytes)
  .toByteArray();
// compare byte arrays ...
```

Reply:
But - maybe I'm just wrong? Like, does the above fail because of unknown fields other than v1/v2 differences?

Reply:
oh i didn't see that! i heard that Go had the same thing, but i didn't look hard enough for a java version

Reply:
much shorter now :) i must have been looking at stackoverflow posts that are outdated now for help on this